File: src/python/counting_sort.py
""" Counting sort in Python """

import random


def counting_sort(arr):
    """ Finding the max element in the list """
    k = max(arr) + 1

    """ Initialing count array of len k with 0's """
    count = [0] * k

    """ Counts frequency of each element """
    for i in arr:
        count[i] += 1

    """ Updates count array to reflect the correct position of each element in the sorted list """
    for i in range(1, k):
        count[i] += count[i - 1]

    """ Initializing result list with 0's """
    result = [0] * len(arr)

    """ Fill result array with the sorted elements"""
    for i in reversed(arr):
        result[count[i] - 1] = i
        count[i] -= 1

    return result


""" Generate a list of n random integers """
n = 10
list = [random.randint(0, 100) for _ in range(n)]

""" Prints original, unsorted list"""
print(f"List: {list}")

""" Sorts list using counting sort algorithm """
sorted_list = counting_sort(list)

""" Prints sorted list """
print(f"Sorted list: {sorted_list}")

File: src/javascript/CountingSort.js
function countingSort(arr) {
    // Find the maximum element in the array to determine the range
    let max = Math.max(...arr);

    // Create a count array to store the count of each element
    let count = new Array(max + 1).fill(0);

    // Count the occurrences of each element in the input array
    for (let i = 0; i < arr.length; i++) {
        count[arr[i]]++;
    }

    // Modify the count array to store the actual position of each element
    for (let i = 1; i <= max; i++) {
        count[i] += count[i - 1];
    }

    // Create a new output array and store the elements in their sorted order
    let output = new Array(arr.length);
    for (let i = arr.length - 1; i >= 0; i--) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }

    return output;
}

const arr = [4, 2, 2, 8, 3, 3, 1];
const sortedArr = countingSort(arr);
console.log("Original array:", arr);
console.log("Sorted array:", sortedArr);
